### PR TITLE
Remove --reset option from provenance.py (force to always write history from scratch)

### LINKED ISSUE #310: Remove provenance --reset option
### Checked for duplicates

No - I haven't checked

### 🧑‍🔬 User Persona(s)

Users, developers

### 💪 Motivation

(Ability to not use) `--reset` has been determined to be an invitation to generate invalid provenance history under certain circumstances, without significant performance benefit.

As such, probably best to remove it entirely

### 📖 Additional Details

_No response_

### Acceptance Criteria

provenance works identically to current behaviour with `--reset` option
provenance throws error if `--reset` flag is provided

### ⚙️ Engineering Details

Breaking change - will require removal of `--reset` from deployment crons.

### PR DESCRIPTION
**Pull request by alexdunnjpl**
## 🗒️ Summary
Removes option to write iterative provenance updates, as this does not improve performance and may result in invalid provenance history in some edge cases. 

Also involves extensive renaming and comments in an attempt to improve documentation of provenance script and speed up future changes.

## ⚙️ Test Data and/or Report
@jimmie do we have any tests that incorporate provenance?  I've tested manually against a minimal dataset.

## ♻️ Related Issues
fixes #310 

If one/some reviewers could sanity-check the comments/renaming, I'd appreciate it - I'm pretty confident that I've captured what's actually going on, but extra eyes never hurt.  Most of the individual commits should be simple-ish to read, taken individually.

### CODE REVIEW

File: support/provenance/provenance.py [hunk 1]
```diff
@@ -102,54 +103,61 @@ def run(
...
+        lidvid_aggregates_by_lid[lid].append(lidvid)
 
-    log.info('   process those with history')
-    count = 0
-    history = {}
-    for lidvids in filter(lambda l: 1 < len(l), aggregates.values()):
-        count += len(lidvids)
+    log.info('   ...determining updated successors for LIDVIDs...')
+    successors_by_lidvid = {}
+    lidvid_aggregates_with_multiple_versions = filter(lambda l: 1 < len(l), lidvid_aggregates_by_lid.values())
```
[the remark this answers is not in the document]
Comment (alexdunnjpl) [reply support/provenance/provenance.py:132]: @sonatype-lift ignore

File: support/provenance/provenance.py [hunk 2]
```diff
@@ -212,9 +223,10 @@ def update_docs(host: HOST, history: {str: str}):
 
 
 if __name__ == '__main__':
-    ap = argparse.ArgumentParser(description='''Update the provenance of products with more than one VID
-
-    The program sweeps through the registry index to find all the lidvids and existing provenance. It then builds up the updates necessary to mark the newly superseded products accordingly.
+    ap = argparse.ArgumentParser(description=f'''
+    Update registry records for non-latest LIDVIDs with up-to-date direct successor metadata ({METADATA_SUCCESSOR_KEY}).
+
```
[the remark this answers is not in the document]
Comment (alexdunnjpl) [reply support/provenance/provenance.py:228]: @sonatype-lift ignore

File: support/provenance/provenance.py [hunk 3]
```diff
@@ -102,54 +103,61 @@ def run(
...
     if updates:
-        update_docs(host, updates)
+        write_updated_docs(host, updates)
 
     log.info('completed CLI processing')
 
 
-def get_historic(provenance: {str: str}, reset: bool) -> {str: str}:  # TODO: populate comment and rename for clarity
-    log.info('starting search for history')
+def get_successors_by_lidvid(extant_lidvids: Iterable[str]) -> Mapping[str, str]:
```
[the remark this answers is not in the document]
Comment (alexdunnjpl) [reply support/provenance/provenance.py:115]: @sonatype-lift ignoreall

File: support/provenance/provenance.py [hunk 4]
```diff
@@ -102,54 +103,61 @@ def run(
...
+    log.info('Generating updated history...')
 
-    log.info('   reduce lidvids to unique lids')
-    lids = sorted({lidvid.split('::')[0] for lidvid in provenance})
+    unique_lids = {lidvid.split('::')[0] for lidvid in extant_lidvids}
 
-    log.info('   aggregate lidvids into lid buckets')
-    aggregates = {lid: [] for lid in lids}
-    for lidvid in provenance: aggregates[lidvid.split('::')[0]].append(lidvid)
+    log.info('   ...binning LIDVIDs by LID...')
```
Comment (nutjob4life) [inline support/provenance/provenance.py:124]: Me being pedantic: use an actual ellipsis character `…` instead of three periods `...` 😅

**Review by nutjob4life — APPROVED** [review]
LGTM 👍
